api/api/catalogue.go

```diff
@@ -0,0 +1,104 @@
+package api
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"net/http"
+	"time"
+	
+	"github.com/gin-gonic/gin"
+
+	"github.com/equinor/seismic-catalogue/api/internal/auth"
+	"github.com/equinor/seismic-catalogue/api/internal/database"
+)
+
+type cubeResponse struct {
+	Country          string `json:"country"          example:"Norway"`
+	Field            string `json:"field"            example:"Volve"`
+	FilenameOnUpload string `json:"filenameOnUpload" example:"myfile.segy"`
+	Url              string `json:"url"              example:"https://account.blob.core.windows.net/cube"`
+}
+
+type Request struct {
+	SasDuration uint `json:"sasDuration" example:"60"`
+}
+
+func makeContainerURL(storageAccount, container string) string {
+	return fmt.Sprintf(
+		"https://%s.blob.core.windows.net/%s",
+		storageAccount,
+		container,
+	)
+}
+
+func signContainerURL(containerURL, sas string) string {
+	return fmt.Sprintf("%s?%s", containerURL, sas)
+}
+
+/* This helper simply aims at hiding the ugly lookup an cast needed to retrieve
+ * values from the gin context
+ */
+func getRolesClaim(ctx *gin.Context) ([]string, error) {
+	roles, exists := ctx.Get("jwtRolesClaim"); 
+	if !exists {
+		return []string{}, errors.New("Expected to find roles claim")
+	}
+	return roles.([]string), nil
+}
+
+func execRequest(
+	ctx          *gin.Context,
+	request      Request,
+	dbConnection database.Adapter,
+	urlSigner    auth.SasFunc,
+) {
+	access, err := getRolesClaim(ctx)
+	if err != nil {
+		ctx.AbortWithError(http.StatusInternalServerError, err)
+		return
+	}
+
+	cubes, err := dbConnection.GetCubes(access)
+	if err != nil {
+		ctx.AbortWithError(http.StatusInternalServerError, err)
+		return
+	}
+
+	duration := time.Duration(request.SasDuration) * time.Minute
+	if duration > time.Duration(24 * time.Hour) {
+		err := errors.New("Maximum duration on sas tokens is 24H")
+		ctx.AbortWithError(http.StatusBadRequest, err)
+		return
+	}
+
+	var response []cubeResponse
+	for _, cube := range cubes {
+		containerURL := makeContainerURL(cube.StorageAccount, cube.Container)
+
+		if request.SasDuration != 0 {
+			sas, err := urlSigner(cube.StorageAccount, cube.Container, duration)
+			if err != nil {
+				ctx.AbortWithError(http.StatusInternalServerError, err)
+				return
+			}
+
+			containerURL = signContainerURL(containerURL, sas)
+		}
+
+		response = append(response, cubeResponse{
+			Country:          cube.Country,
+			Field:            cube.Field,
+			FilenameOnUpload: cube.FilenameOnUpload,
+			Url:              containerURL,
```

Comment:
Btw, we have a very weird design as for our own vds-slice.
Here we return full URL, but our API accepts separately vds and separately sas.

(though with current design users still need to do something more, as this is a link to container, and I currently put vds files into "vds" directory. Unsure if we want to return link to container, or to vds)

Another interesting point:
When I try to use received SAS to read data from vds-slice, I get 403.
But when I use the same service principal in python upload script to generate sas, that sas (with read privileges) can read the data.
Any idea? These SASs work for you?

Reply:
>Btw, we have a very weird design as for our own vds-slice.
Here we return full URL, but our API accepts separately vds and separately sas.

Good point, should sas be it's own entry then? Or does it even make sense that the url and sas is separated, maybe we should change vds-slice?

> and I currently put vds files into "vds" directory. Unsure if we want to return link to container, or to vds)
Then I add /vds

> When I try to use received SAS to read data from vds-slice, I get 403
They broke the signing again 🤦 Now a different version is the only one working